jobqueue_features/tests/test_decorators.py

```diff
@@ -1,41 +1,316 @@
+        self.assertIsInstance(client, Client)
+        controller.delete_cluster(original_cluster.name)
+
+        original_cluster = CustomSLURMCluster(name=_id)
+
+        @on_cluster(cluster_id=_id, cluster=original_cluster)
+        def f():
+            pass
+
+        cluster, client = controller.get_cluster(id_=_id)
+        self.assertEqual(original_cluster, cluster)
+        self.assertEqual(cluster.name, _id)
+        self.assertIsInstance(client, Client)
+        controller.delete_cluster(_id)
+
+    def _test_default_behavior(self, cluster_type):
+        set_default_cluster(cluster_type)
+
+        @on_cluster()
+        def f():
+            pass
+
+        cluster, client = controller.get_cluster()
+        self.assertIsInstance(cluster, cluster_type)
+        self.assertIsInstance(client, Client)
+
+
+class TestTaskDecorator(TestCase):
+    def tearDown(self):
+        controller._close()
+
+    def test_default_behavior(self):
+        cluster_type = CustomSLURMCluster
+        set_default_cluster(cluster_type)
+        _value = 123
 
         @on_cluster()
         @task()
-        def example_function():
-            return "test"
+        def f():
+            return _value
+
+        self._basic_tests(cluster_type)
+
+    def test_task_by_id_string(self):
+        _id = "test1"
+        cluster_type = CustomSLURMCluster
+        original_cluster = cluster_type(name=_id)
+
+        @on_cluster(cluster=original_cluster)
+        @task(cluster_id=_id)
+        def f():
+            pass
+
+        self._basic_tests(cluster_type=cluster_type, id_=_id)
+        controller.delete_cluster(_id)
+
+    def test_task_by_cluster_name(self):
+        _id = "test1"
+        cluster_type = CustomSLURMCluster
+        original_cluster = cluster_type(name=_id)
+
+        @on_cluster(cluster=original_cluster)
+        @task(cluster_id=original_cluster.name)
+        def f():
+            pass
+
+        self._basic_tests(cluster_type=cluster_type, id_=original_cluster.name)
+        controller.delete_cluster(original_cluster.name)
+
+    def test_task_by_cluster(self):
+        cluster_type = CustomSLURMCluster
+        original_cluster = cluster_type()
+
+        @on_cluster(cluster=original_cluster)
+        @task(cluster=original_cluster)
+        def f():
+            pass
+
+        self._basic_tests(cluster_type=cluster_type, id_=original_cluster.name)
+        controller.delete_cluster(original_cluster.name)
+
+    def test_task_by_cluster_local(self):
+        _id = "test1"
+        cluster_type = LocalCluster
+        original_cluster = cluster_type()
+
+        with self.assertRaises(ClusterException) as ctx:
+            @on_cluster(cluster=original_cluster, cluster_id=_id)
+            @task(cluster=original_cluster)
+            def f():
+                pass
+        self.assertEqual("'cluster_id' argument is required for LocalCluster.",
+                         ctx.exception.message)
```

Comment:
From https://stackoverflow.com/a/43453668 you should probably use `str(ctx.exception)`

Reply:
Done